### ISSUE TITLE
Notify when updates are available (Mac)

### ISSUE THREAD

**Issue by deeplow** [freedomofpress/dangerzone#189]
Notify the user when another Mac release is available.

**Comment by deeplow** [comment]
ProtonVPN has a built-in update process. Maybe for mac we'd want to take inspiration from their implementation. Or maybe this is something build-in to macOS. Will have to investigate.

<img width="321" alt="proton" src="https://user-images.githubusercontent.com/47065258/185387932-573f1aa9-48c1-4531-8b62-aee80c512352.png">

**Comment by keywordnew** [comment]
I suggest that disabling update checks should prevent a built-in update feature from creating network requests, in addition to not displaying notifications that an update is available.

This flow would account for the usecase where users prefer using a package manager, like Brew, to handle updates and rely on firewalls to control outbound network requests. A built-in update feature that triggers the firewall, despite the update feature being disabled, would be avoidably noisy and may create concern.

**Comment by apyrgio** [comment]
@deeplow mentioned something interesting, that I was not aware of. There's a class of software, like Sparkle Project, that deals solely with safely updating an already installed application. So, before diving into how can Dangerzone notify users for updates, it would help laying down what's the current state with these update frameworks:

# Update Frameworks

## Sparkle Project

[Sparkle project](https://sparkle-project.org) seems to be the de-facto update framework in MacOS. Integrators in practice need to care about two things: creating a proper `Appcast.xml` file on the server-side, and calling the Sparkle code from the client-side. These are covered in the project's [documentation](https://sparkle-project.org/documentation/).

The client-side part is not very straight-forward, since Sparkle is written in Objective-C. Thankfully, there are others who have ventured into this before: https://fman.io/blog/codesigning-and-automatic-updates-for-pyqt-apps/

The server-side part is also not very straight-forward. For integrators that use GitHub releases (like us), this issue may be of help: https://github.com/sparkle-project/Sparkle/issues/648

The Windows platform is not covered by Sparkle itself, but there are other projects, such as [WinSparkle](https://winsparkle.org/), that follow a similar approach. I see that there's a [Python library (`pywinsparkle`)](https://pypi.org/project/pywinsparkle/) for interacting with WinSparkle, so this may alleviate some pains.

Note that the Sparkle project is not a silver bullet. Development missteps can happen, and users can be left without updates. Here's an [example issue](https://github.com/sparkle-project/Sparkle/issues/345) that showcases this. 

## The Update Framework

[The Update Framework](https://theupdateframework.io/) is a graduated CNCF project hosted by Linux Foundation. It's based on the [Thandy](https://chromium.googlesource.com/chromium/src.git/+/master/docs/updater/protocol_3_1.md) updater for Tor. It's [not widely adopted](https://github.com/sparkle-project/Sparkle/issues/345), but some of its adopters are high-profile, and it has passed security audits.

It's more of a [specification](https://github.com/sparkle-project/Sparkle/issues/345) and less of a software project, although a well-maintained [reference implementation](https://github.com/sparkle-project/Sparkle/issues/345) in Python exists.

Regardless of whether we use it, knowing about the [threat vectors](https://theupdateframework.io/security/) that it's protecting against is very important.

## Other Projects

Qt has some updater framework as well: https://doc.qt.io/qtinstallerframework/ifw-updates.html
Google Chrome has it's own updater framework: https://chromium.googlesource.com/chromium/src.git/+/master/docs/updater/protocol_3_1.md
Keepass rolls out its own way to update: https://github.com/keepassxreboot/keepassxc/blob/develop/src/updatecheck/UpdateChecker.cpp
[PyUpdater](https://github.com/Digital-Sapphire/PyUpdater) was another popular updater project for Python, but is now archived.

**Comment by apyrgio** [comment]
Integrating with an update framework will certainly take a bit of work, especially given that we target both Windows and MacOS systems. In the meantime though, we can't have releases out without including at least a notification channel, since staying behind on updates has a huge negative impact on the users' safety.

So, we can keep this issue scoped to just the notification part, and punt the problem of installing updates to a subsequent release.

Here's a suggestion on how we can proceed.

# Design overview

## First run

_We detect it's the first time Dangerzone runs because the `settings["updater"]` key is not set._

Add the following keys in our `settings.json` file.

```
{
    "updater": {
        "check": None,  # Whether to check for updates or not. None means that the user has not decided yet, and is the default.
        "latest_version": "0.4.2",  # The latest version that the Dangerzone updater has detected. By default it's the current version.
        "errors: [],  # The list of errors that the updater has encountered. By default it's empty.
    }
}
```

* If on Linux, make `"check": False`, since we normally have other update channels for these releases.
* TODO: Detect Homebrew installations, to avoid unnecessary checks.

## Second run

_We detect it's the second time Dangerzone runs because `settings["updater"]["check"] is None`._

Before starting up the main window, show this window:

* Title: Dangerzone Updater
* Body:

  > Do you want to be notified about subsequent releases?
  > 
  > By clicking on "Yes", Dangerzone will check GitHub for new releases on startup. By clicking on "No", Dangerzone will make no network requests and won't prompt you for releases.
  > 
  > If you prefer another way of getting notified about new releases, we suggest adding to your RSS reader our Atom feed: https://github.com/freedomofpress/dangerzone/releases.atom. For more information about updates, check our wiki page: https://github.com/freedomofpress/dangerzone/wiki/Updates
* Buttons:
   - Yes: Store `settings["updater"]["check"] = True`.
   - No: Store `settings["updater"]["check"] = False`

## Subsequent runs

_We perform the following only if `settings["updater"]["check"] == True`._

1. Spawn a new thread so that we don't block the main window.
2. Hit the GitHub releases API and get the [latest release](https://api.github.com/repos/freedomofpress/dangerzone/releases/latest).
3. Check if the latest release matches `settings["updater"]["latest_version"]`. If yes, return.
4. Show this new window:
    * Title: "Dangerzone v0.5.0 has been released"
    * Body:
    
       > A new version of Dangerzone has been released. You can download it from https://dangerzone.rocks.
       >
       > Changelog:
       >
       > (Show changelog rendered from Markdown in a scrollable text box)
       >
       > By clicking on "OK", you won't be reminded again. You can click on "Reminder me later" to be reminded on the next time Dangerzone starts.

    * Buttons:
       - OK: Store the latest version in `settings["updater"]["latest_version"]`, clear the `settings["updater"]["errors"]` key.
       - Remind me later: Do not store the latest version, clear the `settings["updater"]["errors"]` key.

## Error handling

_We trigger error handling when the updater thread encounters an error (either due to an HTTPS failure or a Python exception) and does not complete successfully._

1. Append the error message and a timestamp in the settings:

    ```settings["updater"]["errors"].append((datetime.datetime.utcnow(), error_msg))```
    
2. Check if more than 3 errors have been recorded, and a week has passed since the first one. If not, return.
3. Show a window:
    * Title: "Dangerzone Updater"
    * Body:
       
       > Something went wrong while checking for Dangerzone updates:
       >
       > (Show the latest error message in a scrollable, copyable text box)
       >
       > You are strongly advised to visit https://dangerzone.rocks and check for new updates manually, or consult our wiki page (https://github.com/freedomofpress/dangerzone/wiki/Updates) for common causes of errors . Alternatively, you can click on "Disable updates", if you are in an air-gapped environment and have another way of learning about updates
       
   * Buttons:
      - OK: Clear the `settings["updater"]["errors"]` key.
      - Disable updates: Store `settings["updater"]["check"] = False` and clear the `settings["updater"]["errors"]` key.

**Comment by apyrgio** [comment]
# Key Benefits

1. The above approach future-proofs Dangerzone against API changes or bugs in the update check process, by asking users to manually visit https://dangerzone.rocks.
2. If we want to draw the attention of users to immediately install a release, we can do so in the release body, which we will show in a pop-up window.
3. If we are aware of issues that prevent updates, we can add them in the wiki page that we show in the error popup. Wiki pages are not versioned, so we can add useful info even after a release.

# Security Considerations

Because this approach does not download binaries / auto-updates, it **does not add any more security issues** than the existing, manual way of installing updates. These issues have to do with a compromised/malicous GitHub service, and are the following:

1. GitHub pages can alter the contents of our main site (https://dangerzone.rocks)
2. GitHub releases can serve an older, vulnerable version of Dangerzone, instead of a new update.
3. GitHub releases can serve a malicious binary (requires a joint operation from a malicious CA as well, for extra legitimacy). 
4. GitHub releases can silently drop updates.
5. GitHub releases can know which users download Dangerzone updates.
6. Network attackers can know that a user has Dangerzone installed (because we ask the user to visit https://dangerzone.rocks)

A good update framework would probably defend against 1,2,3. This is not to say that our users are currently unprotected, since 1-4 can be detected by the general public and the developers (unless GitHub specifically targets an individual, but that's another story).

# Usability Considerations

1. If a user initially does not want to enable update checks, but then changes their mind, they need to visit our wiki page to find out how.
2. We do not have an update story for users that only use the Dangerzone CLI. A good assumption is that they are on Linux, so they auto-update.
3. We don't have a way to remind users about a new Dangerzone release, say after a week from the previous notification, unless they click on "Remind me later".

**Comment by eloquence** [comment]
Thanks much for the detailed research and thoughts :) 
- Agree with keeping installation of updates out of scope.
- I'd suggest pointing to https://fosstodon.org/@dangerzone (RSS: https://fosstodon.org/@dangerzone.rss) instead of the GitHub feed as a way for end users to stay up-to-date. We're able to include non-release announcements there, and it's generally more human-friendly. (If we migrate servers at a future time, Mastodon supports account migration with redirects.)
- UX-wise, I'd suggest considering an inline notification (e.g., a green bar "Dangerzone update available [Details]") at the top of the window) over a pop-up, as a way to minimize user disruption. Especially if we check updates in the background after the application has started, initiating a pop-up while the user is in the middle of a task flow can be disruptive.

**Comment by apyrgio** [comment]
> ### Fallback mechanism
> 
> > Hit the GitHub releases API and get the [latest release](https://api.github.com/repos/freedomofpress/dangerzone/releases/latest).
> 
> Fingers crossed that that API never ever changes. We should probably have our own fallback URL (even if generally inactive), just in case the user ever gets anything other than 200 OK. Considering Github outages, we may not want to rely fully on them.

I had thought of something similar as well. If I get what you're saying, you talk about serving a file, probably with a different schema, under https://dangerzone.rocks, right? It makes sense, but:

* https://dangerzone.rocks is hosted on GitHub Pages,
* the uptime of GitHub API requests from the start of the year till May is ~99% (39 hours of outages in 5 months)
* we'd need to store the same info in two places

I think this approach will add complexity with diminishing returns.

> ### Strict version increase checking
> 
> > Check if the latest release matches settings["updater"]["latest_version"]. If yes, return.
> 
> > GitHub releases can serve an older, vulnerable version of Dangerzone, instead of a new update.
> 
> Let's probably check if it's superior. Just as a sanity check?

Doesn't matter much, because the attacker can say that they serve a later release, but in reality the binary may be an older one. There is no way to detect this unless you download the binary.

> ### Check interval?
> 
> We probably don't need to check every time the app starts. Maybe we a `last_check` date setting that only asks if it's hasn't in X amount of time. Otherwise, we are providing github with an uneeded level of "telemtry" of app startups ([kind of apple-style](https://arstechnica.com/gadgets/2020/11/mac-certificate-check-stokes-fears-apple-logs-every-app-you-run/)).

Pretty reasonable, I agree.

> ### Way to change setting
> 
> > If a user initially does not want to enable update checks, but then changes their mind, they need to visit our wiki page to find out how.
> 
> Why not just add a discrete button to the bottom of the "select suspicious docs" window? ![deeee](https://user-images.githubusercontent.com/47065258/247980041-7e9c1852-bbc4-49d1-bb23-d2ab6dddb42d.png)

Oh, I hadn't considered touching our UI, before we have a UX discussion! I think what you're suggesting would tie really nice with @eloquence's suggestion for an inline notification in our GUI. I really like that

> ### Other notes
> 
> > GitHub pages can alter the contents of our main site (https://dangerzone.rocks/)
> 
> I think it's under FPF's control at the moment. But still any push will be mirrored so that probably still applies.

Mentioned it above; it's served by GitHub Pages actually.
